### PR TITLE
Adding periodic boundary condition support to tobac

### PR DESCRIPTION
**Pull request by galexsky**
Extensive update of tobac capabilities and procedures by myself and @freemansw1 . Many substantial changes to code, such as:

- Inclusion of the vertical dimension for 3D feature detection and tracking, now accounting for a potentially non-uniform dz
- Modification to 3D segmentation allowing "box seeding" of user-set size in addition to default "column seeding" method
- Feature detection, segmentation, and tracking now able to address datasets with periodic boundary conditions (PBCs)
- Addition of several new utility functions and movement of some such functions from other modules to utils
- Optimization, simplification, bug fixes, and other streamlining of overall code base
- Improved code documentation

### CODE REVIEW

**Comment by freemansw1** [comment]
A few notes: 

- Given the magnitude of the changes, I've created a new branch for this, called `dev-3D-PBC`. 
- I contributed to this substantially, so I cannot review the PR. Instead, I've requested @JuliaKukulies and @w-k-jones .
- Some documentation improvements are also contained in here. I genuinely forgot that those were in here as well, so we are happy to remove them. 
- #124 hasn't been merged yet, but note that coverage (of feature detection, tracking, and segmentation) has increased substantially. Once we get #124 merged, hopefully the codecov bot will run here. It still needs improvement, especially on the feature detection (and analysis) side, but this is overall better.
- There are still some TODO items, including running 3D feature detection on a subset of the points. However, @galexsky and I decided that it was better to go ahead and get this in as a PR so that we can work on these new features collaboratively. 

I think I speak for both @galexsky and myself when we say that we're happy for any comments on this. This project has been a long time coming, so there are likely some bugs scattered about. Thanks also to @grleung and @smsaleeb for beta testing this.

**Comment by JuliaKukulies** [comment]
All right, thank you for your great efforts @freemansw1 and @galexsky! In my opinion, we can keep the documentation bit in this PR, as this is the part that does not take a lot of effort to review and then we have those improvements at one sweep!

As you can see, I started the review, which looks a bit unsystematic at the moment, but I thought it would be good to directly comment on minor comments and questions as well, while I am going through the code step-by-step to understand and review your additions. It will take some time to go through all changes, as I am also testing your additions with some real data. So far these extra tests look good, no problems with the automatic tests in my local clone and documentation rendering looks also fine!

**Comment by freemansw1** [comment]
I'll get to addressing your comments on this ~next week. I know Alex is quite busy finishing up for his defense at this point.

**Comment by freemansw1** [comment]
As I'm beginning to work on this, I'm realizing more and more that we should really split out the documentation stuff. 3D and PBC is a lot harder to split out, but the docs stuff should be fairly easy. If @galexsky doesn't object, I'll create a new PR for just documentation stuff and see if I can get documentation stuff out of this PR.

**Comment by JuliaKukulies** [comment]
OK, that would be great if it is not too much work you @freemansw1 @galexsky !

File: doc/feature_detection_base_out_vars.csv [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿Variable Name,Description,Units,Type
+frame,Frame/time/file number; starts from 0 and increments by 1 to N times. ,n/a,int64
+idx,"Feature number within that frame; starts at 1, increments by 1 to the number of features for each frame, and resets to 1 when the frame increments",n/a,int64
+hdim_1,"First horizontal dimension in grid point space (typically, although not always, N/S or y space)",Number of grid points,float64
+hdim_2,"Second horizontal dimension in grid point space (typically, although not always, E/W or x space)",Number of grid points,float64
+num,Number of grid points that are within the threshold of this feature,Number of grid points,int64
+threshold_value,Maximum threshold value reached by the feature,Units of the input feature,int64(?)
```
[the remark this answers is not in the document]
Comment (freemansw1) [reply doc/feature_detection_base_out_vars.csv]: These changes have now been moved to #150 .

File: tobac/centerofgravity.py [hunk 2]
```diff
@@ -10,8 +10,9 @@ def calculate_cog(tracks, mass, mask):
                 cube of quantity (need coordinates 'time', 'geopotential_height','projection_x_coordinate' and 'projection_y_coordinate')
     mask:       iris.cube.Cube
                 cube containing mask (int > where belonging to cloud volume, 0 everywhere else )
-    Output:
-    tracks_out  pandas.DataFrame
+    Returns
+    -------
+    pandas.DataFrame
```
[the remark this answers is not in the document]
Comment (freemansw1) [reply tobac/centerofgravity.py]: I've reverted these changes in light of #138

File: tobac/feature_detection.py [hunk 3]
```diff
@@ -212,92 +348,382 @@ def feature_detection_threshold(
...
+    x_min = 0
+    x_max = labels.shape[2] - 1
+
+    # deal with PBCs
+    # all options that involve dealing with periodic boundaries
+    pbc_options = ["hdim_1", "hdim_2", "both"]
+
+    # we need to deal with PBCs in some way.
+    if PBC_flag in pbc_options:
+        #
```
[the remark this answers is not in the document]
Comment (freemansw1) [reply tobac/feature_detection.py:395]: This is a really interesting idea, and certainly one that could be tackled. I think we could move this to a new `pbc_utils` module or something inside the `utils` module. We could also move some of the functions from segmentation in there, and would make readability easier. I'll work on this a bit and see what we can do.

File: tobac/tracking.py [hunk 4]
```diff
@@ -23,33 +27,64 @@ def linking_trackpy(
...
                   time resolution of tracked features
     dxy:          float
-                  grid spacing of input data
+                  Horizontal grid spacing of input data
+    dz: float
+        Constant vertical grid spacing of input data. If None,
+        uses `vertical_dim` to get vertical location.
     memory        int
                   number of output timesteps features allowed to vanish for to be still considered tracked
     subnetwork_size int
```
[the remark this answers is not in the document]
Comment (freemansw1) [reply tobac/tracking.py:54]: I believe that this is being addressed in #138, so I will leave this as-is for now.

File: tobac/segmentation.py [hunk 5]
```diff
@@ -209,6 +1077,9 @@ def segmentation(
     method="watershed",
     max_distance=None,
     vertical_coord="auto",
+    PBC_flag="none",
+    seed_3D_flag="column",
+    seed_3D_size=5,
```
[the remark this answers is not in the document]
Comment (freemansw1) [reply tobac/segmentation.py:1082]: Yes- the intent was for it to be able to be specified by the user. I've resolved this bug.

**Comment by freemansw1** [comment]
Okay, I've done the following: 

1.  the branch we are merging into, `dev-3d-pbc`, has been updated to the latest `dev`. 
2. As discussed, documentation changes have largely been pulled out to #150. There are some updated docstrings in here, but I will wait until #138 is merged before trying to get those all cleaned up and merged.
3. I've addressed several of the changes raised by @JuliaKukulies . I'm going to work on separating out the PBC functions into a separate utility class, but I want to do a PR addressing #122 et al. first.

**Comment by freemansw1** [comment]
I'm back to working on this now that we're very close to v1.4.0. I've got the latest changes from RC_v1.5.0 merged in, although there will be some more work to do once #190 is merged in. 

I'm not ready for a re-review as there's a lot more to do, but I will ping when I am.

**Comment by freemansw1** [comment]
Okay, I've updated this code for `RC_V1.5.0`. One thing that I'm struggling with: I cannot get our workaround for predictive tracking working with 3D data. For now, I've switched tests to use `random`, but that is clearly not ideal. @snilsn any ideas? I know that this will be fixed once https://github.com/soft-matter/trackpy/pull/710 is merged, but without an ETA there, it would be much better to have a workaround. 

@snilsn if you're able/willing/interested in helping, I've put the "base case" in a gist here: https://gist.github.com/freemansw1/a917e0d823524c6bb44eee458e218a02 .

**Comment by snilsn** [comment]
I could get `df_link_iter` to produce an output by setting `pos_columns` to ['z', 'y', 'x'] manually, but then the prediction doesn't work. This could mean that guessing the position columns isn't working correctly for 3d inputs within `df_link_iter`, which would make our workaround useless. `trackpy.utils.guess_pos_columns()` is used for thi. Seems to work fine

One idea I had was specifyng an 3d `initial_guess_vels` in `NearestVelocityPredict`, but that isn't helping.

Did you find any examples for trackpy that include tracking 3d features using predictors @freemansw1?

**Comment by snilsn** [comment]
I think there is a solution (at least theoretically). It's possible to set the values for the interpolator manually:

```
from scipy.interpolate import NearestNDInterpolator
import numpy as np
import pandas as pd
import trackpy

test_df = pd.DataFrame({'z': [0,1,2,3], 'y':[0,1,2,3], 
'x': [0,1,2,3], 'frame': [0,1,2,3]})

features_list = [frame for i, frame in test_df.groupby("frame", sort=True)]

initial_guess_pos = np.asarray([[0, 0, 0]])
initial_guess_vel = np.asarray([[1, 1, 1]])

pred = trackpy.predict.NearestVelocityPredict(span=1)
pred.interpolator = NearestNDInterpolator(initial_guess_pos, initial_guess_vel)

traj_iter = trackpy.link_df_iter(features_list, 
                                 search_range=0.1, 
                                 t_column='frame', 
                                 predictor=pred.predict)
pd.concat(traj_iter)
```

This code links all features into one particle despite the small search range, i. e. the prediction is happening. 

The problem is, that the value for the velocity isn't adjusted automatically. This would mean that we would have to loop through all the frames in pairs, interpolate the velocity from the last frame and set a new `pred.interpolator` at each step, do the linking for the two frames and combine everything into one dataframe at the end.  

It's probably possible, but a little ridiculous.

**Comment by freemansw1** [comment]
I've now merged in the latest v1.5.0 changes, most importantly the utils breakout (#191). I'm in the process of thinking how best to do this. I'm becoming inclined to break out the 3D from the PBC changes, despite the pain that that will likely cause for me. As of now, this PR is extremely difficult to review due to its size and scope, with around 5K LOC changed. 

I'm going to hold off asking for re-reviews until I resolve this.

**Comment by freemansw1** [comment]
It appears that, because #209 was merged and contains this commit history, GitHub has tagged this as merged as well. I'll work on a new PR for the PBC changes.

**Comment by w-k-jones** [comment]
That's a slightly annoying feature on Github! Let us know when the new PR is ready and I'll review it asap

**Pull request merged**